fdbclient/MultiVersionClientControl.actor.cpp

```diff
@@ -0,0 +1,193 @@
+/*
+ * MultiVersionClientControl.actor.cpp
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2013-2021 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "fdbclient/MultiVersionClientControl.actor.h"
+#include "fdbclient/Schemas.h"
+#include "fdbclient/NativeAPI.actor.h"
+#include "fdbclient/ManagementAPI.actor.h"
+#include "fdbrpc/IAsyncFile.h"
+#include "flow/Platform.h"
+
+#include <algorithm>
+#include <string>
+#include <stdio.h>
+
+#include "flow/actorcompiler.h" // This must be the last #include.
+
+constexpr size_t CLIENTLIB_CHUNK_SIZE = 8192;
+constexpr size_t CLIENTLIB_TRANSACTION_SIZE = CLIENTLIB_CHUNK_SIZE * 32;
+
+struct ClientLibBinaryInfo {
+	size_t totalBytes;
+	size_t chunkCnt;
+};
+
+ACTOR Future<Void> uploadClientLibBinary(Database db,
+                                         StringRef libFilePath,
+                                         KeyRef chunkKeyPrefix,
+                                         ClientLibBinaryInfo* binInfo) {
+
+	state size_t fileOffset = 0;
+	state size_t chunkNo = 0;
+
+	state Reference<IAsyncFile> fClientLib = wait(IAsyncFileSystem::filesystem()->open(
+	    libFilePath.toString(), IAsyncFile::OPEN_READONLY | IAsyncFile::OPEN_UNCACHED | IAsyncFile::OPEN_NO_AIO, 0));
+
+	loop {
+		state Arena arena;
+		state StringRef buf = makeString(CLIENTLIB_TRANSACTION_SIZE, arena);
+		state int bytesRead = wait(fClientLib->read(mutateString(buf), CLIENTLIB_TRANSACTION_SIZE, fileOffset));
+		fileOffset += bytesRead;
+		if (bytesRead <= 0) {
+			break;
+		}
+
+		state Transaction tr(db);
+		state size_t firstChunkNo = chunkNo;
+		loop {
+			try {
+				tr.setOption(FDBTransactionOptions::ACCESS_SYSTEM_KEYS);
+				size_t bufferOffset = 0;
+				chunkNo = firstChunkNo;
+				while (bufferOffset < bytesRead) {
+					size_t chunkLen = std::min(CLIENTLIB_CHUNK_SIZE, bytesRead - bufferOffset);
+					KeyRef chunkKey = chunkKeyPrefix.withSuffix(format("%06zu", chunkNo), arena);
+					chunkNo++;
+					tr.set(chunkKey, ValueRef(mutateString(buf) + bufferOffset, chunkLen));
+					bufferOffset += chunkLen;
+				}
+
+				wait(success(tr.getReadVersion()));
```

Comment:
`getReadVersion` is implied by your first call to read or commit, so you shouldn't need to explicitly call it here.

Reply:
The main motivations I can think of for calling it explicitly would be if you want to know the value of the read version or if you want to time how long the GRV call took by itself.

Reply:
Call removed